lib/Sema/CSSolver.cpp

```diff
@@ -652,6 +652,18 @@ bool ConstraintSystem::tryTypeVariableBindings(
         type = type->reconstituteSugar(/*recursive=*/false);
       }
 
+      // Instead of binding collection types directly let's try to
+      // bind using temporary type variables substituted for element
+      // types, that's going to ensure that subtype relationship is
+      // always preserved.
+      if (isConcreteCollectionType(type)) {
```

Comment:
This should probably be an `else if` clause of the `if` above, since we don't want to look back at a type that we just opened to type variables.